### PR TITLE
finish error rewrite

### LINKED ISSUE #83: Error handling
Improve error handling.

- [ ] More detailed lifetime extension validation results
- [ ] More detailed capabilities extension validation
- [ ] Make framing `verify` functions return errors

### PR DESCRIPTION
**Pull request by franziskuskiefer**
This PR finishes the error rewrite. It also introduces a new type of errors.
Now the following is allowed as well.
```Rust
implement_error! {
    pub enum CredentialError {
        Simple {
            UnsupportedCredentialType = "Unsupported credential type",
        }
        Complex {
            ConfigError(ConfigError) = "See `ConfigError` for details.",
            CryptoError(CryptoError) = "See `CryptoError` for details.",
        }
    }
}
```

fixes #83

### CODE REVIEW

File: src/group/errors.rs [hunk 1]
```diff
@@ -28,28 +28,32 @@ implement_error! {
...
+            TreeHashMismatch =
+                "The tree hash computed does not match the one in the GroupInfo.",
+            ConfirmationTagMismatch =
+                "The computed confirmation tag does not match the expected one.",
+            InvalidGroupInfoSignature =
+                "The signature on the GroupInfo is not valid.",
+            GroupInfoDecryptionFailure =
+                "Unable to decrypt the GroupInfo",
+            DuplicateRatchetTreeExtension =
+                "Found more than one ratchet tree extension in the Welcome message.",
```
Comment (raphaelrobert) [inline src/group/errors.rs:47]: "Found more a duplicate ratchet tree extension in the Welcome message."

File: src/group/errors.rs [hunk 2]
```diff
@@ -28,28 +28,32 @@ implement_error! {
...
-            "An unknown error occurred.",
+        Simple {
+            CiphersuiteMismatch =
+                "Ciphersuites in the Welcome message and the corresponding key package bundle don't match.",
+            JoinerSecretNotFound =
+                "No joiner secret found in the Welcome message.",
+            MissingRatchetTree =
+                "No ratchet tree available to build initial tree after receiving a Welcome message",
+            TreeHashMismatch =
+                "The tree hash computed does not match the one in the GroupInfo.",
```
Comment (raphaelrobert) [inline src/group/errors.rs:39]: "The computed tree hash does not match the one in the GroupInfo."

**Review by raphaelrobert — CHANGES_REQUESTED** [review]
Thanks! Just a few nits.

I think we will still need follow-ups to surface more of the library errors/document the unwraps.

**Review by kkohbrok — APPROVED** [review]
Looks good, thanks! (Nothing to add to @raphaelrobert's review.)

**Comment by franziskuskiefer** [comment]
> Thanks! Just a few nits.
> 
> I think we will still need follow-ups to surface more of the library errors/document the unwraps.

Yeah we'll definitely need to better with error handling. But now we have the tools in place and all existing errors are handled in the same way

**Review by raphaelrobert — APPROVED** [review]

**Pull request merged**